ruby/ql/lib/codeql/ruby/security/UnsafeCodeConstructionQuery.qll

```diff
@@ -0,0 +1,49 @@
+/**
+ * Provides a taint-tracking configuration for reasoning about code
+ * constructed from library input vulnerabilities.
+ *
+ * Note, for performance reasons: only import this file if `Configuration` is needed,
+ * otherwise `UnsafeCodeConstructionCustomizations` should be imported instead.
+ */
+
+import codeql.ruby.DataFlow
+import UnsafeCodeConstructionCustomizations::UnsafeCodeConstruction
+private import codeql.ruby.TaintTracking
+private import codeql.ruby.dataflow.BarrierGuards
+
+/**
+ * A taint-tracking configuration for detecting code constructed from library input vulnerabilities.
+ */
+class Configuration extends TaintTracking::Configuration {
+  Configuration() { this = "UnsafeShellCommandConstruction" }
+
+  override predicate isSource(DataFlow::Node source) { source instanceof Source }
+
+  override predicate isSink(DataFlow::Node sink) { sink instanceof Sink }
+
+  override predicate isSanitizer(DataFlow::Node node) {
+    node instanceof StringConstCompareBarrier or
+    node instanceof StringConstArrayInclusionCallBarrier
+  }
+
+  // override to require the path doesn't have unmatched return steps
+  override DataFlow::FlowFeature getAFeature() {
+    result instanceof DataFlow::FeatureHasSourceCallContext
```

Comment:
Just a question - I'm not familiar with `FlowFeature`s - why is predicate necessary for this particular flow configuration?

Reply:
It prevents flow out of functions that weren't called as part of the path.   
I have a test-case for it here: https://github.com/github/codeql/blob/44213f0144fdd54bb679ca48d68b28dcf820f7a8/ruby/ql/test/query-tests/security/cwe-078/UnsafeShellCommandConstruction/impl/unsafeShell.rb#L37-L44 

In that example the `x` parameter in `id()` is correctly identified as a source.  
However, if that value flows out of the `id()` function (into e.g. `thisIsSafe()`), then it is no longer a valid source.  

This flow-feature blocks that kind of flow for returns that doesn't have a matching call.